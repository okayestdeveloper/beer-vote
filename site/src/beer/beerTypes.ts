import { IProfile } from './../hooks/useProfile';

export const BEER_COLOR_MAP: Array<string> = [
  '#FFE699',
  '#FFD878',
  '#FFCA5A',
  '#FFBF42',
  '#FBB123',
  '#F8A600',
  '#F39C00',
  '#EA8F00',
  '#E58500',
  '#DE7C00',
  '#D77200',
  '#CF6900',
  '#CB6200',
  '#C35900',
  '#BB5100',
  '#B54C00',
  '#B04500',
  '#A63E00',
  '#A13700',
  '#9B3200',
  '#952D00',
  '#8E2900',
  '#882300',
  '#821E00',
  '#7B1A00',
  '#771900',
  '#701400',
  '#6A0E00',
  '#660D00',
  '#5E0B00',
  '#5A0A02',
  '#600903',
  '#520907',
  '#4C0505',
  '#470606',
  '#440607',
  '#3F0708',
  '#3B0607',
  '#3A070B',
  '#36080A',
  '#000000',
  '#000000',
  '#000000',
  '#000000',
  '#000000',
  '#000000',
  '#000000',
  '#000000',
  '#000000',
  '#000000',
];

export enum EBeerStatus {
  VOTING = 'voting',
  DRAFT = 'draft',
  AVAILABLE = 'available',
  BREWING = 'brewing',
}

export enum EServingFormat {
  BOTTLE_12_OZ = '12oz bottle',
  BOTTLE_22_OZ = '22oz bottle',
  BOTTLE_500_ML = '500ml bottle',
  BOTTLE_750_ML = '750ml bottle',
  CAN_12_OZ = '12oz can',
  CAN_16_OZ = '16oz can',
  DRAUGHT = 'draught',
}

export interface IBeer {
  id: string;
  ABV?: number;
  maxABV: number;
  minABV: number;
  description?: string;
  name: string;
  style: string;
  servingFormats?: EServingFormat[];
  tags?: string[];
  status: EBeerStatus;
  votes: number;
  srm: number;
}

// Props
export interface IBeerListProps {
  beers: IBeer[];
  upvoteBeer: (beer: IBeer, profile: IProfile) => void;
}

export interface IBeerCardProps {
  beer: IBeer;
  upvote: (beer: IBeer, profile: IProfile) => void;
}
