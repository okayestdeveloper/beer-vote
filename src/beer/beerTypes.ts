export enum EBeerStatus {
  VOTING = 'voting',
  DRAFT = 'draft',
  AVAILABLE = 'available',
  BREWING = 'brewing',
}

export enum EServingFormat {
  BOTTLE_12_OZ = '12oz bottle',
  BOTTLE_22_OZ = '22oz bottle',
  BOTTLE_500_ML = '500ml bottle',
  BOTTLE_750_ML = '750ml bottle',
  CAN_12_OZ = '12oz can',
  CAN_16_OZ = '16oz can',
  DRAUGHT = 'draught',
}

export interface IBeer {
  id: string;
  ABV?: number;
  maxABV: number;
  minABV: number;
  description?: string;
  name: string;
  style: string;
  servingFormats?: EServingFormat[];
  tags?: string[];
  status: EBeerStatus;
  votes: number;
}

// Props
export interface IBeerListProps {
  beers: IBeer[];
  upvoteBeer: (beer: IBeer) => void;
}

export interface IBeerCardProps {
  beer: IBeer;
  upvote: (beer: IBeer) => void;
}
