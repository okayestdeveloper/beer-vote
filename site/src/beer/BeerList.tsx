import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';

import BeerCard from './BeerCard';
import { RootState } from '../store/rootReducer';
import { IBeer } from './beerTypes';
import { fetchBeers } from './beerSlice';

function sortByVotes(beerA: IBeer, beerB: IBeer): number {
  return beerB.votes - beerA.votes;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    listContainer: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'flex-start',
      flexWrap: 'wrap',
      maxWidth: 1056,
      margin: '0 auto',
    },
  }),
);

function BeerList(): React.ReactElement {
  const dispatch = useDispatch();

  const beers = useSelector((state: RootState) => {
    const sorted = state.beers.beers.slice();
    sorted.sort(sortByVotes);
    return sorted;
  });

  useEffect(() => {
    dispatch(fetchBeers());
  }, []);

  // todo: somehow re-sort after upvote
  const classes = useStyles();
  return (
    <div className={classes.listContainer}>
      {beers.map((beer) => (
        <BeerCard beer={beer} key={beer.id} />
      ))}
    </div>
  );
}

export default BeerList;
