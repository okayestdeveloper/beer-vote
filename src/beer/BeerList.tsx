import React from 'react';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';

import BeerCard from './BeerCard';
import { IBeerListProps } from './beerTypes';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    listContainer: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'flex-start',
      flexWrap: 'wrap',
      maxWidth: 1056,
      margin: '0 auto',
    },
  }),
);

const BeerList: React.FC<IBeerListProps> = ({ beers, upvoteBeer }) => {
  const classes = useStyles();
  return (
    <div className={classes.listContainer}>
      {beers.map((beer) => (
        <BeerCard beer={beer} upvote={upvoteBeer} key={beer.id} />
      ))}
    </div>
  );
};

export default BeerList;
