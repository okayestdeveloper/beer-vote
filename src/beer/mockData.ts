import { IBeer, EServingFormat, EBeerStatus } from './beerTypes';

const beers: IBeer[] = [
  {
    id: 'beer1',
    maxABV: 7.0,
    minABV: 6.0,
    description: 'A dark chocolate milk stout.',
    name: 'Valkyrie',
    style: 'Chocolate Stout',
    servingFormats: [EServingFormat.BOTTLE_12_OZ, EServingFormat.CAN_12_OZ],
    tags: ['dark', 'chocolate', 'rich'],
    status: EBeerStatus.VOTING,
    votes: 543,
  },
  {
    id: 'beer2',
    ABV: 5.5,
    maxABV: 4.0,
    minABV: 6.0,
    description: 'An American Pale Ale with tropical, fruity hops.',
    name: 'Seeker',
    style: 'Juicy Pale Ale',
    servingFormats: [EServingFormat.DRAUGHT],
    tags: ['juicy', 'bright', 'refreshing'],
    status: EBeerStatus.AVAILABLE,
    votes: 32,
  },
];

export const getBeers = (): IBeer[] => {
  return JSON.parse(JSON.stringify(beers));
};
