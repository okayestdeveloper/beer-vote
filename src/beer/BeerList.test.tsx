import React from 'react';
import {
  cleanup,
  render,
  fireEvent,
  RenderResult,
} from '@testing-library/react';
import renderer from 'react-test-renderer';

import { IBeer, IBeerListProps } from './beerTypes';
import { getBeers } from './mockData';
import BeerList from './BeerList';

describe(`BeerList rendering`, () => {
  let beers: IBeer[];

  beforeEach(() => {
    beers = getBeers();
  });

  it(`should render a list of beers`, () => {
    const tree = renderer.create(
      <BeerList beers={beers} upvoteBeer={jest.fn()} />,
    );
    expect(tree).toMatchSnapshot();
  });
});

describe(`BeeList props`, () => {
  let beerList: RenderResult;
  let beers: IBeer[];
  let mockUpvote: jest.Mock;

  function renderList(args?: Partial<IBeerListProps>): RenderResult {
    const defaultProps: IBeerListProps = {
      beers,
      upvoteBeer: () => {},
    };

    const props = { ...defaultProps, ...args };
    return render(<BeerList {...props} />);
  }

  beforeEach(() => {
    beers = getBeers();
    mockUpvote = jest.fn();
    beerList = renderList({ upvoteBeer: mockUpvote });
  });

  afterEach(cleanup);

  it(`should call upvoteBeer when the event is passed up from the child`, () => {
    const { getAllByTestId } = beerList;
    const btns = getAllByTestId('upvoteButton');
    fireEvent.click(btns[0]);
    expect(mockUpvote).toHaveBeenCalledWith(beers[0]);
  });
});
