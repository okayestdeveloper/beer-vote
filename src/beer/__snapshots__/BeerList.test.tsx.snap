// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerList rendering should render a list of beers 1`] = `
<div>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-beerCard-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardHeader-root"
    >
      <div
        className="MuiCardHeader-content"
      >
        <span
          className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Valkyrie
        </span>
        <span
          className="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          Chocolate Stout
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
      >
        A dark chocolate milk stout.
      </p>
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
      >
        6
        % - 
        7
        % ABV
      </p>
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
      >
        543
         Votes
      </p>
      <button
        aria-label="upvote"
        className="MuiButtonBase-root MuiIconButton-root"
        data-testid="upvoteButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root makeStyles-upvoteIcon-2"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 8V4l8 8-8 8v-4H4V8z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-beerCard-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardHeader-root"
    >
      <div
        className="MuiCardHeader-content"
      >
        <span
          className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Seeker
        </span>
        <span
          className="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          Juicy Pale Ale
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
      >
        An American Pale Ale with tropical, fruity hops.
      </p>
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
      >
        5.5
        %
      </p>
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary"
      >
        32
         Votes
      </p>
      <button
        aria-label="upvote"
        className="MuiButtonBase-root MuiIconButton-root"
        data-testid="upvoteButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root makeStyles-upvoteIcon-2"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 8V4l8 8-8 8v-4H4V8z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
